File: 1971. Find if Path Exists in Graph.cpp
class Solution
{
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination)
    {
        vector<bool> visited(n, false);
        vector<vector<int>> graph = buildGraph(n, edges);

        dfsRecursion(source, graph, visited);

        return visited[destination];
    }

    vector<vector<int>> buildGraph(int n, vector<vector<int>>& edges)
    {
        vector<vector<int>> graph(n, vector<int>());

        for (auto edge : edges)
        {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        return graph;
    }

    void bfs(int start, vector<vector<int>>& graph, vector<bool>& visited)
    {
        queue<int> q;
        q.push(start);
        visited[start] = true;

        while (not q.empty())
        {
            int v = q.front();
            q.pop();

            for (int u : graph[v])
            {
                if (not visited[u])
                {
                    visited[u] = true;
                    q.push(u);
                }
            }
        }
    }

    void dfs(int start, vector<vector<int>>& graph, vector<bool>& visited)
    {
        stack<int> s;
        s.push(start);
        visited[start] = true;

        while (not s.empty())
        {
            int v = s.top();
            s.pop();

            for (int u : graph[v])
            {
                if (not visited[u])
                {
                    visited[u] = true;
                    s.push(u);
                }
            }
        }
    }

    void dfsRecursion(int v, vector<vector<int>>& graph, vector<bool>& visited)
    {
        visited[v] = true;

        for (int u : graph[v])
        {
            if (not visited[u])
            {
                dfsRecursion(u, graph, visited);
            }
        }
    }
};

File: 1971. Find if Path Exists in Graph.cs
public class Solution
{
    public bool ValidPath(int n, int[][] edges, int source, int destination)
    {
        var graph = new List<IList<int>>();

        for(int i = 0; i < n; ++i)
        {
            graph.Add(new List<int>());
        }

        foreach (var edge in edges)
        {
            graph[edge[0]].Add(edge[1]);
            graph[edge[1]].Add(edge[0]);
        }

        var isUsed = new HashSet<int>();

        Dfs(source, graph, isUsed);

        return isUsed.Contains(destination);
    }

    private void Dfs(int v, IList<IList<int>> graph, HashSet<int> isUsed)
    {
        isUsed.Add(v);

        foreach (var u in graph[v])
        {
            if (!isUsed.Contains(u))
            {
                Dfs(u, graph, isUsed);
            }
        }
    }
}